### ISSUE TITLE
Concurrency issues with Artifacts

### ISSUE THREAD

**Issue by jwikman** [microsoft/navcontainerhelper#1088]
**Describe the issue**
We've got a lot of pipelines running in parallell on the same (hosted) build server.
Earlier, when we had different images published for each locale and version, docker daemon handled concurrency when several pipelines requested to download the same image.

Today I've been running a lot of pipelines with the new Artifact approach, and I've seen some strange issues that seems to be related to concurrency.

I've only been able to reproduce one of them by now, I will add more if I'm able to repro later.

### Repro
1. Clear your bcartifacts cache to be sure that artifacts are downloaded
1. Also clear you docker image cache if you've got any images called 'myimage' :)
1. Open PowerShell ISE (as admin) and open two tabs (two separate PowerShell sessions) with a new editor window in each tab.

4. In the first tab you paste this:
```
$credential = New-Object pscredential 'super', (ConvertTo-SecureString -String "P@ssword" -AsPlainText -Force)
$parameters = @{
    "accept_eula"              = $true
    "accept_outdated"          = $true
    "containerName"            = 'test2'
    "artifactUrl"              = "https://bcartifacts.azureedge.net/sandbox/16.3.14085.14363/se"
    "imageName"                = 'myimage'
    "auth"                     = "NavUserPassword"
    "Credential"               = $credential
    "dns"                      = '8.8.8.8'
    "alwaysPull"               = $true
    "updateHosts"              = $true
    "includeAL"                = $true
    "includeCSide"             = $false
    "doNotExportObjectsToText" = $true
    "shortcuts"                = "none"
    "myScripts"                = @(@{'MainLoop.ps1' = 'while ($true) { start-sleep -seconds 10 }' })
    "doNotCheckHealth"         = $true
    "enableTaskScheduler"      = $false
}

New-NavContainer @parameters
```
5. In the second tab you paste this (The only differences is the containerName):
```
$credential = New-Object pscredential 'super', (ConvertTo-SecureString -String "P@ssword" -AsPlainText -Force)
$parameters = @{
    "accept_eula"              = $true
    "accept_outdated"          = $true
    "containerName"            = 'test1'
    "artifactUrl"              = "https://bcartifacts.azureedge.net/sandbox/16.3.14085.14363/se"
    "imageName"                = 'myimage'
    "auth"                     = "NavUserPassword"
    "Credential"               = $credential
    "dns"                      = '8.8.8.8'
    "alwaysPull"               = $true
    "updateHosts"              = $true
    "includeAL"                = $true
    "includeCSide"             = $false
    "doNotExportObjectsToText" = $true
    "shortcuts"                = "none"
    "myScripts"                = @(@{'MainLoop.ps1' = 'while ($true) { start-sleep -seconds 10 }' })
    "doNotCheckHealth"         = $true
    "enableTaskScheduler"      = $false
}

New-NavContainer @parameters
```

6. Now we will try to mimic how two pipelines are starting at pretty much the same time by running both scripts directly after eachother.

### Problem

One of the scripts will be first to complete the download of the artifacts needed and starts to unzip and rename the tmp folder ('D:\bcartifacts.cache\sandbox\16.3.14085.14363\se-tmp' in my case) to the destination name ('D:\bcartifacts.cache\sandbox\16.3.14085.14363\se'). But the other get stuck at a loop:
```` 
Renaming tmp folder
Trying to get exclusive access to d:\bcartifacts.cache\sandbox\16.3.14085.14363\se-tmp - waiting 10 seconds and retrying...
Renaming tmp folder
Trying to get exclusive access to d:\bcartifacts.cache\sandbox\16.3.14085.14363\se-tmp - waiting 10 seconds and retrying...
Renaming tmp folder
Trying to get exclusive access to d:\bcartifacts.cache\sandbox\16.3.14085.14363\se-tmp - waiting 10 seconds and retrying...
Renaming tmp folder
Trying to get exclusive access to d:\bcartifacts.cache\sandbox\16.3.14085.14363\se-tmp - waiting 10 seconds and retrying...
Renaming tmp folder
Trying to get exclusive access to d:\bcartifacts.cache\sandbox\16.3.14085.14363\se-tmp - waiting 10 seconds and retrying...
Renaming tmp folder
Trying to get exclusive access to d:\bcartifacts.cache\sandbox\16.3.14085.14363\se-tmp - waiting 10 seconds and retrying...
Renaming tmp folder
Trying to get exclusive access to d:\bcartifacts.cache\sandbox\16.3.14085.14363\se-tmp - waiting 10 seconds and retrying...
Renaming tmp folder
Trying to get exclusive access to d:\bcartifacts.cache\sandbox\16.3.14085.14363\se-tmp - waiting 10 seconds and retrying...
Renaming tmp folder
Trying to get exclusive access to d:\bcartifacts.cache\sandbox\16.3.14085.14363\se-tmp - waiting 10 seconds and retrying...
````

We've got a lot of pipelines scheduled every night with several countries in each pipeline. Every time the latest version change, the pipelines will try to download artifacts and build new images - and they will do it at the same time - so we need some mechanism to handle these scenarios.

Will try to repro the other issues as well, I think that there was a problem when both processes where trying to build the same image at the same time...

**Comment by jwikman** [comment]
Just as I posted I got a new error, but caused by the same issue as above:
````
Downloading C:\Users\nabjohwik\AppData\Local\Temp\2ed55bcd-2503-4fc7-83bf-d47d54bf183e.zip
Unpacking application artifact to tmp folder
Remove-Item : Cannot remove item D:\bcartifacts.cache\sandbox\16.3.14085.14363\us-tmp\database\database.bak: The process cannot access the file 'database.bak' because it is being used by another process.
At D:\VSCode\Git\NAB\NAB DevOps Tools\PowerShell\navcontainerhelper\Misc\Download-Artifacts.ps1:80 char:21
+ ...              Remove-Item -Path "$appArtifactPath-tmp" -Recurse -Force
+                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : WriteError: (database.bak:FileInfo) [Remove-Item], IOException
    + FullyQualifiedErrorId : RemoveFileSystemItemIOError,Microsoft.PowerShell.Commands.RemoveItemCommand
````

This happens if one process is renaming the tmp-folder at the same time as the other process unzips to the tmp-folder...

**Comment by jwikman** [comment]
Next error, happens when artifacts are already downloaded and cached. One process has a lock on the lastused file, so the other errors out.

```` 
Removing C:\ProgramData\NavContainerHelper\Extensions\test1
ArtifactUrl and ImageName specified
Set-Content : The process cannot access the file 'D:\bcartifacts.cache\sandbox\16.3.14085.14363\us\lastused' because it is being used by another process.
At D:\VSCode\Git\NAB\NAB DevOps Tools\PowerShell\navcontainerhelper\Misc\Download-Artifacts.ps1:98 char:9
+         Set-Content -Path (Join-Path $appArtifactPath 'lastused') -Va ...
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Set-Content], IOException
    + FullyQualifiedErrorId : System.IO.IOException,Microsoft.PowerShell.Commands.SetContentCommand
````

**Comment by freddydk** [comment]
One fix you can do right now is to add this line to the place you load the containerhelper:

$navContainerHelperConfig.bcartifactsCacheFolder = "bcartifacts.<agentname>"

You probably already guessed what it does.
Yes, you will not have shared cache - but it should fix these issues, while I think about a solution.

**Comment by jwikman** [comment]
Unfortunately I don't think it will help completely, we're just pushing the errors in front of us. 

Both processes uses the same temp folder when building images ('c:\$TMP$-myimage-sandbox-16.3.14085.14363-us' in my sample below) and that gave me new errors. Now both processes failed

````
Building image myimage:sandbox-16.3.14085.14363-us based on https://bcartifacts.azureedge.net/sandbox/16.3.14085.14363/us
Pulling latest image mcr.microsoft.com/dynamicsnav:10.0.18363.900-generic
10.0.18363.900-generic: Pulling from dynamicsnav
Digest: sha256:08ea7b63826c65fa291a6cab87476fad36a8f05ea7cfc7968a1d88732153f800
Generic Tag: 0.1.0.7
Container OS Version: 10.0.18363.900 (1909)
Host OS Version: 10.0.18363.900 (1909)
Copying Platform Artifacts
Copy Database
Copy Extensions
c:\$TMP$-myimage-sandbox-16.3.14085.14363-us
Sending build context to Docker daemon  785.8MB

Step 1/6 : FROM mcr.microsoft.com/dynamicsnav:10.0.18363.900-generic
 ---> 08e3365bf962
Step 2/6 : ENV DatabaseServer=localhost DatabaseInstance=SQLEXPRESS DatabaseName=CRONUS IsBcSandbox=Y artifactUrl=https://bcartifacts.azureedge.net/sandbox/16.3.14085.14363/us
 ---> Using cache
 ---> a83fe5412c89
Step 3/6 : COPY my /run/
docker : COPY failed: CreateFile \\?\D:\Docker\tmp\docker-builder806394834\my: The system cannot find the file specified.
At D:\VSCode\Git\NAB\NAB DevOps Tools\PowerShell\navcontainerhelper\ContainerHandling\New-NavImage.ps1:294 char:1
+ docker build --isolation=$isolation --memory $memory --tag $imageName ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (COPY failed: Cr...file specified.:String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError
````

````
Building image myimage:sandbox-16.3.14085.14363-us based on https://bcartifacts.azureedge.net/sandbox/16.3.14085.14363/us
Pulling latest image mcr.microsoft.com/dynamicsnav:10.0.18363.900-generic
10.0.18363.900-generic: Pulling from dynamicsnav
Generic Tag: 0.1.0.7
Container OS Version: 10.0.18363.900 (1909)
Host OS Version: 10.0.18363.900 (1909)
Remove-Item : Cannot remove item C:\$TMP$-myimage-sandbox-16.3.14085.14363-us\NAVDVD\WebClient\Microsoft Dynamics NAV\160\Web Client\WebPublish\wwwroot\Resources: The directory is not empty.
At D:\VSCode\Git\NAB\NAB DevOps Tools\PowerShell\navcontainerhelper\ContainerHandling\New-NavImage.ps1:152 char:9
+         Remove-Item $buildFolder -Force -Recurse
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : WriteError: (Resources:DirectoryInfo) [Remove-Item], IOException
    + FullyQualifiedErrorId : RemoveFileSystemItemIOError,Microsoft.PowerShell.Commands.RemoveItemCommand

````

**Comment by freddydk** [comment]
Correct - I know what is going on - will see what I can do

**Comment by jwikman** [comment]
Thanks, appreciate this.

Nothing of this did happen when I was changing our pipeline framework to use artifacts in the dev branch, since when I was running the tests I was not running several pipelines at the same time.

So now when I updated the master branch and all repos with pipelines to use this I get errors all over the place. And the pipelines that are scheduled tonight against next minor will probably fail, since that is always new and will generate new images...

**Comment by freddydk** [comment]
Yes - and this will happen every time I update the generic images or every time you uptake a new version if we don't get this fixed

**Comment by jwikman** [comment]
Exactly.
And we get new versions for next minor every night.... So tomorrow we'll have a bunch of bugs created because next minor builds failed

**Comment by freddydk** [comment]
Only if you are running the same country version in multiple agents - at the same time, but yes.

**Comment by jwikman** [comment]
They are all scheduled at the same time, with several agents picking them up. And they all runs against pretty much the same countries... :)

Let me know if you need any more input or if I can test anything out.

**Comment by freddydk** [comment]
Testing concurrency...
![image](https://user-images.githubusercontent.com/10775043/86944962-bc595a80-c148-11ea-9a5b-f3277c6e86f1.png)

**Comment by freddydk** [comment]
![image](https://user-images.githubusercontent.com/10775043/86945088-e448be00-c148-11ea-8974-0c8885db5120.png)

**Comment by freddydk** [comment]
![image](https://user-images.githubusercontent.com/10775043/86945174-04787d00-c149-11ea-9660-3f6e3a1c5808.png)

**Comment by freddydk** [comment]
![image](https://user-images.githubusercontent.com/10775043/86945500-60db9c80-c149-11ea-9203-57bcf6247c06.png)

#1 catched up....

**Comment by freddydk** [comment]
![image](https://user-images.githubusercontent.com/10775043/86945887-e4958900-c149-11ea-8179-5015ba7b4a94.png)

**Comment by freddydk** [comment]
![image](https://user-images.githubusercontent.com/10775043/86945998-0989fc00-c14a-11ea-8e77-51d68eee8165.png)

**Comment by freddydk** [comment]
![image](https://user-images.githubusercontent.com/10775043/86946434-a8165d00-c14a-11ea-9641-4438b1768a61.png)

It seems like I have something that is way more stable wrt. concurrency now.

**Comment by jwikman** [comment]
That indeed looks promising. 🙂

Do you want me to test anything, or will you push it out?

**Comment by freddydk** [comment]
If you want to try (before I can publish a new containerhelper), you can download a new New-NavImage.ps1 here https://www.dropbox.com/s/qh8x4eh4cqdi40y/New-NavImage.ps1?dl=1 and a new Download-Artifacts,ps1 here https://www.dropbox.com/s/l5mxnk05l6tt5fw/Download-Artifacts.ps1?dl=1
download, right-click, select properties and unblock file and override the existing file in the containerhelper module.

**Comment by freddydk** [comment]
Current implementation might fail if two pipelines do the same thing within 100 nanoseconds (I am using [datetime]::now.ticks for my temp foldername (and build folder)
and when renaming - if the folder exists I assume that somebody else have been there (as it wasn't there before downloading)

**Comment by jwikman** [comment]
Downloaded and about to test.

**Comment by jwikman** [comment]
I think we can rule out the possibility of having two temp folders with the same name now. :)

Several pipelines creating the same image at the same time seems to be possible now (even if it is a waste of resources, since we only need 1 image) 

It might also be the waste of disk space, since both images seems to be created - but only one is used. But I can live with this for now.
![image](https://user-images.githubusercontent.com/23032645/86949499-05aca880-c14f-11ea-91be-62f4e50db01f.png)

The folder rename _should_ be instantly, since you are renaming on the same drive. So that _should_ be fine.

I will push this out on our pipelines later tonight and try to break it ;-)


**Bottom line:**
This seems to work - at least good enough for now, good job - Thanks! :)

(We might need some Mutex logic to make it completely foolproof, with one mutex per ArtifactURL or something like that...)

**Comment by freddydk** [comment]
On your build server - you could run do Docker System Prune to get rid of dangling images and dead containers.
I will probably come up with a cleanup function at some point in time.
Maybe Prune followed by deletion of all autobuild images, where the artifacts haven't been used for x days.

**Comment by jwikman** [comment]
Yes, I'm having cleanup scheduled once a week - but now I might need to schedule that more often until we've got a more efficient way of keeping down storage needs :D

Are you updating the lastused file in artifacts also when an existing image is used? In that case your suggestion sounds perfect.

**Comment by freddydk** [comment]
lastused is updated on the artifacts when the image is used, yes. (not the platform, but that doesn't matter), so basically, this script:

```
docker system prune -f
Flush-ContainerHelperCache -cache bcartifacts -keepDays 7
$myimagename = "myimage"
docker images --format "{{.Repository}}:{{.Tag}}" | where-object { $_.Split(':')[0] -eq $myimagename -and (!(Test-Path (Join-Path $navContainerHelperConfig.bcartifactsCacheFolder ($_.Split(':')[1].Replace('-','\'))))) }  | % { docker rmi $_ -f }
```
Would clean up artifacts and docker images auto-created with the name myimage that hasn't been used for the last 7 days.

**Comment by jwikman** [comment]
Cool, thanks - will try it out scheduled a bit more frequently

I did not encounter any more concurrency errors this night, so fix seems to work.

Only errors I got was pipeline timeouts - several pipelines took more than 6 hours to complete, since they are slowing each other down by downloading artifacts and  creating images at the same time. I upped the timeout to 10 hours for now, so I can go on vacation. :)
I will think about how the concurrency performance could be improved later on, might flip in a PR if I find a good way of doing this.

**Comment by jwikman** [comment]
Check out what you think about my mutex solution of this issue in PR #1095 

It gives significant time gains and should also save some disk space. And I assume that those gains are linear to the number of parallell jobs...

**Comment by jwikman** [comment]
Now I've played around a bit with my fix, running six pipelines in parallell with both same and different localizations - some with up to 20 different artifacts to create containers for.

All pipelines seems to behave very good, waiting nicely in queue for their turn to check cached artifacts and images. The container creations all are faster now, since the hardware resources are not throttling in the same way as before.
I also see a lot less dangling images (none), so it allocates less disk space. (every dangling image took over 3gb, if I interpret things correctly...)

This approach also lets you revert the temp folder name change you did last night, I've seen a few cases where the temp folders are being left - probably caused by canceled pipeline runs...
If we have fixed temp folder names (as before) and there is a temp folder for that artifact already, you can safely delete that one and be sure that it was leftovers from a dead thread.

Anyway, I'm off for vacation tomorrow. But now I'm confident that the pipelines will work nicely for my colleagues when I'm not available. (I just might need to expand the data disks on the build server)

**Comment by jwikman** [comment]
Freddy, Inspired by your cleanup script above I started to think about what kind of BCArtifacts we are using in our pipelines. We are running pipelines for both Sandbox and OnPrem. For OnPrem we're using a fixed version for each pipeline, the CU version that particular customer are running OnPrem. For Sandbox pipelines we always run against latest version - for current version, next minor and next major when the release in approaching.

So the cleanup taks for OnPrem images are quite easy. Your script will do it, if they have been used in the last X days - keep the image. 

For Sandbox images I simply check if the image still is the latest one, if not - delete the image (and it's BCArtifact related Cache)

I ended up with below script that now is scheduled every night through Windows Task Scheduler. Use or abuse (or ignore) as you want. :)

```
Write-Host @"
# Remove:
#  - all stopped containers
#  - all networks not used by at least one container
#  - all dangling images
#  - all build cache
"@ -ForegroundColor Green
docker system prune -f

Write-Host "# Clear bcartifacts.cache that has not been used for 14 days" -ForegroundColor Green
Install-Module navcontainerhelper -Force
Import-Module -name navcontainerhelper -DisableNameChecking
Flush-ContainerHelperCache -cache bcartifacts -keepDays 14

Write-Host "# Remove created images that does not exists in bcartifacts.cache" -ForegroundColor Green
$myimagename = "nabbcimage"
docker images --format "{{.Repository}}:{{.Tag}}" |`
  where-object { $_.Split(':')[0] -eq $myimagename -and `
  (!(Test-Path (Join-Path $navContainerHelperConfig.bcartifactsCacheFolder ($_.Split(':')[1].Replace('-','\'))))) }  | % { 
    Write-Host "Image $_ has no cache in '$($navContainerHelperConfig.bcartifactsCacheFolder)' - remove image" -ForegroundColor Red
    docker rmi $_ -f
}



Write-Host "# Remove images and cache that are not 'latest'" -ForegroundColor Green
$Images = docker images --format "{{.Repository}}:{{.Tag}}" | where-object { $_.Split(':')[0] -eq $myimagename } 
$Images | ForEach-Object {

    $imageName = $_ 
    Write-Host "Check if image $imageName should be deleted"

    $ArtifactParts = $imageName.Split(':')[1].Split('-')

    $ImageArtifactUrl = Get-BCContainerArtifactUrl -containerName $imageName
    $ImageArtifactUri = [Uri]::new($ImageArtifactUrl)

    $Artifact = @{
        Type           = $ArtifactParts[0]
        Version        = [version]::new($ArtifactParts[1])
        Country        = $ArtifactParts[2]
        SasToken       = $ImageArtifactUri.Query
        StorageAccount = $ImageArtifactUri.Host.Split('.')[0]
    }

    $removeImage = $false
    if ($Artifact.Type.ToLower() -eq "onprem") {
        Write-Host "Image $ImageName is used for OnPrem, keep this image"
    }
    else {
        try {
            $Params = @{
                StorageAccount = $Artifact.StorageAccount
                type           = $Artifact.Type
                country        = $Artifact.Country
                version        = $Artifact.Version.Major
                select         = "Latest"
            }
            if ($Artifact.SasToken -ne "") {
                $Params += @{
                    sasToken = $Artifact.SasToken
                }
            }
                    $LatestArtifactUrl = Get-BCArtifactUrl @Params
            $LatestArtifactUri = [Uri]::new($LatestArtifactUrl)
            $LatestVersion = [Version]::new($LatestArtifactUri.AbsolutePath.Split('/')[2])
            if ($LatestVersion -eq $Artifact.Version) {
                Write-Host "Image $ImageName has the latest available version, keep this image"
            }
            else {
                Write-Host "Image $ImageName has an older version ($($Artifact.Version)) than the latest available version for this major ($LatestVersion), remove this image" -ForegroundColor Yellow
                $removeImage = $true
            }
        }
        catch {
            Write-Host "Could not retrieve the latest Artifact Url corresponding to Image $imageName, leave the image to other cleanup tasks" -ForegroundColor Red
        }
    }

    if ($removeImage) {
    
        try {
            Write-Host "Delete image $imageName" -ForegroundColor Yellow
            docker rmi $imageName
        }
        catch {
            Write-Host "Image $imageName could not be deleted, probably used by an existing container" -ForegroundColor Red
        }
        try {
            $BCArtifactCachePath = Join-Path $navContainerHelperConfig.bcartifactsCacheFolder ("$($Artifact.Type)\$($Artifact.Version)\$($Artifact.Country)")
            Write-Host "Remove cache folder '$BCArtifactCachePath'" -ForegroundColor Yellow
            if (Test-Path $BCArtifactCachePath) {
                Remove-Item -Path $BCArtifactCachePath -Recurse -Force
            }

            $VersionCachePath = Join-Path $navContainerHelperConfig.bcartifactsCacheFolder ("$($Artifact.Type)\$($Artifact.Version)")
            if (Test-Path $VersionCachePath ) {
                if ((Get-ChildItem -Path $VersionCachePath -Exclude "platform") -eq $null) {
                    Write-Host "Last cached country version, remove version folder '$VersionCachePath'" -ForegroundColor Yellow
                    Remove-Item -Path $VersionCachePath -Recurse -Force
                }
            }
        }
        catch {
            Write-Host "Cache folder '$BCArtifactCachePath' could not be deleted, just leave it then." -ForegroundColor Red
        }
    }
}
```

**Comment by freddydk** [comment]
Shipped in 0.7.0.14
